### PR TITLE
vdk-jupyter: make run before deploy optional

### PR DESCRIPTION
**Pull request by duyguHsnHsn**
What: Added a checkbox button that determines whether the job will be run once before deployment.
Screenshot 2023-08-21 at 10 19 53

Why: Currently, the job is always running but in some use cases the users might not want it. 

Testing: modified the jest tests
Signed-off-by: Duygu Hasan [hduygu@vmware.com](mailto:hduygu@vmware.com)

### CODE REVIEW

**Comment by antoniivanov** [comment]
As usual for new UI elements, please provide a screenshot

File: projects/vdk-plugins/vdk-jupyter/vdk-jupyterlab-extension/src/components/DeployJob.tsx [hunk 1]
```diff
@@ -48,53 +49,82 @@ export default class DeployJobDialog extends Component<IJobFullProps> {
...
+      }
     } catch (error) {
       await showErrorMessage(
         'Encountered an error when deploying the job. Error:',
         error,
         [Dialog.okButton()]
       );
-    }
-  }
+    }}
```
Comment (yonitoo) [inline projects/vdk-plugins/vdk-jupyter/vdk-jupyterlab-extension/src/components/DeployJob.tsx:129]: nit: fix these stacked parentheses

**Review by yonitoo — APPROVED** [review]
LGTM!

**Comment by murphp15** [comment]
Does this run it locally? 
why would you want to run before deployment? 
If you wanted to run it before deployment would you not just run it using the start button?

**Comment by duyguHsnHsn** [comment]
> Does this run it locally? why would you want to run before deployment? If you wanted to run it before deployment would you not just run it using the start button?

Mainly to ensure that the deployed job will be green. Yes, it is always an option for them to run it manually like that this option is added just for convenience. 
Imagine doing some config.ini changes, not changes in the code u might actually forget about running it again, this works both as reminder and if the run fails u get an error message and the job is not deployed, but if it is green u get no update on the run but directly the jo bis deployed

File: projects/vdk-plugins/vdk-jupyter/vdk-jupyterlab-extension/src/__tests__/show-dialog-deploy.spec.tsx [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+      const mockResult = { button: { accept: true }, value: true };
+      (showDialog as jest.Mock).mockResolvedValueOnce(mockResult);
+      // Mock the jobRunRequest function
+      (jobRunRequest as jest.Mock).mockResolvedValueOnce({
+        message: 'Job completed successfully!',
+        status: true
+      });
+    });
+
+    it('should show a dialog with the Create Deployment title and \
```
Comment (antoniivanov) [inline projects/vdk-plugins/vdk-jupyter/vdk-jupyterlab-extension/src/__tests__/show-dialog-deploy.spec.tsx:38]: Ensure that all new functionalities are covered with tests including error cases and edge cases (For example, what happens when the checkbox is checked/unchecked?, verify the by default checkbox is clicked  )

File: projects/vdk-plugins/vdk-jupyter/vdk-jupyterlab-extension/src/components/DeployJob.tsx [hunk 3]
```diff
@@ -48,46 +49,77 @@ export default class DeployJobDialog extends Component<IJobFullProps> {
...
+            type="checkbox"
+            name="deployRun"
+            id="deployRun"
+            className="jp-vdk-checkbox"
+            onChange={handleCheckboxChange}
+          />
+          <label className="checkboxLabel" htmlFor="deployRun">
+            Run before deployment
+          </label>
+        </div>
```
Comment (antoniivanov) [inline projects/vdk-plugins/vdk-jupyter/vdk-jupyterlab-extension/src/components/DeployJob.tsx:89]: Consider further component modularization (see Separation of Concerns). The checkbox for 'Run before deployment' can be a separate reusable component.

**Review by antoniivanov — CHANGES_REQUESTED** [review]

File: projects/vdk-plugins/vdk-jupyter/vdk-jupyterlab-extension/src/components/DeployJob.tsx [hunk 4]
```diff
@@ -48,46 +49,77 @@ export default class DeployJobDialog extends Component<IJobFullProps> {
...
+          jobTeam={jobData.get(VdkOption.TEAM) || ''}
+        />
+        <div>
+          <input
+            type="checkbox"
+            name="deployRun"
+            id="deployRun"
+            className="jp-vdk-checkbox"
+            onChange={handleCheckboxChange}
+          />
```
Comment (antoniivanov) [inline projects/vdk-plugins/vdk-jupyter/vdk-jupyterlab-extension/src/components/DeployJob.tsx:85]: The checkbox should be checked by default. That is running the job prior to deployment is opt-out (not opt-in) . This is it encourage developers to always test and run their code prior to deployment.

File: projects/vdk-plugins/vdk-jupyter/vdk-jupyterlab-extension/src/components/DeployJob.tsx [hunk 5]
```diff
@@ -48,46 +49,77 @@ export default class DeployJobDialog extends Component<IJobFullProps> {
...
+          <label className="checkboxLabel" htmlFor="deployRun">
+            Run before deployment
```
Comment (antoniivanov) [inline projects/vdk-plugins/vdk-jupyter/vdk-jupyterlab-extension/src/components/DeployJob.tsx:87]:
```suggestion
            Run data job before deployment
```

File: projects/vdk-plugins/vdk-jupyter/vdk-jupyterlab-extension/src/components/DeployJob.tsx (hunk 4 above)
Comment (antoniivanov) [reply projects/vdk-plugins/vdk-jupyter/vdk-jupyterlab-extension/src/components/DeployJob.tsx:85]: Because if you deploy using Notebook, you've developed it using notebooks. 

The running before deployment address some of the biggest issues with using notebooks - out of order cell executions, lack of reproducability/replicability . There's a nice podcast where a guy explain those (https://towardsdatascience.com/the-case-against-the-jupyter-notebook-d4da17e97243) 

So by running it before deployment with vdk run, we ensure the job would run the same way and finish as expected before deployment and after deployment in the cloud.

File: projects/vdk-plugins/vdk-jupyter/vdk-jupyterlab-extension/src/components/DeployJob.tsx [hunk 6]
```diff
@@ -48,53 +49,82 @@ export default class DeployJobDialog extends Component<IJobFullProps> {
       </>
     );
   }
+
+  handleCheckboxChange = (event: React.ChangeEvent<HTMLInputElement>) => {
```
[the remark this answers is not in the document]
Comment (duyguHsnHsn) [reply projects/vdk-plugins/vdk-jupyter/vdk-jupyterlab-extension/src/components/DeployJob.tsx:53]: Yes, one of the with the window element was a previous implementation (i didn't realise it was left there since it does not do anything). i have deleted it

File: projects/vdk-plugins/vdk-jupyter/vdk-jupyterlab-extension/src/components/DeployJob.tsx (hunk 4 above)
Comment (duyguHsnHsn) [reply projects/vdk-plugins/vdk-jupyter/vdk-jupyterlab-extension/src/components/DeployJob.tsx]: changed it to checked by default

File: projects/vdk-plugins/vdk-jupyter/vdk-jupyterlab-extension/src/__tests__/show-dialog-deploy.spec.tsx (hunk 2 above)
Comment (duyguHsnHsn) [reply projects/vdk-plugins/vdk-jupyter/vdk-jupyterlab-extension/src/__tests__/show-dialog-deploy.spec.tsx:40]: ready

**Review by antoniivanov — APPROVED** [review]
